### PR TITLE
Feature/remove unused properties

### PR DESCRIPTION
**Pull request by nmck257**
## What's changed?

Added a new `RemoveUnusedProperties` maven recipe

## What's your motivation?

Seems useful for certain hygiene tasks

## Anything in particular you'd like reviewers to focus on?

My `parentHasProperty` method -- I *think* that it:
- Can expect ~100% `PomCache` hits and not actually cause new downloads
- Does *not* "poison" the pom cache / have weird side effects by resolving this fake, stubbed version of this project pom
- Is the simplest impl I can think of

...but open to other ideas :)

## Any additional context

Running this with the default `.+` matcher can be a little over-eager, particularly for properties which support custom plugins, but the recipe seems quite practical for specific use cases (eg cleaning invalid `.+\\.version` properties for a Spring Boot project)

### Checklist
- [X] I've added unit tests to cover both positive and negative cases
- [X] I've read and applied the [recipe conventions and best practices](https://docs.openrewrite.org/authoring-recipes/recipe-conventions-and-best-practices)
- [X] I've used the IntelliJ IDEA auto-formatter on affected files

### CODE REVIEW

File: rewrite-maven/src/test/java/org/openrewrite/maven/RemoveUnusedPropertiesTest.java [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+                    <artifactId>module2</artifactId>
+                  </project>
+                  """
+              )
+            )
+          )
+        );
+    }
+
+}
```
[the remark this answers is not in the document]
Comment (nmck257) [reply rewrite-maven/src/test/java/org/openrewrite/maven/RemoveUnusedPropertiesTest.java:416]: gh seems to disagree with the robot about whether there's actually a change being suggested here

**Comment by timtebeek** [comment]
Great to see; I know this has come up as a request quite a bit over the past couple years, but it's always been tricky to get right. Have you given any though to [properties used in resource filtering](https://maven.apache.org/plugins/maven-resources-plugin/examples/filter.html) for instance? I feel that might complicate matters here.

**Comment by nmck257** [comment]
zero thought into that! but at a glance, extending the scanner to check `src/main/resources` should be feasible? I can take a look next week

File: rewrite-maven/src/main/java/org/openrewrite/maven/RemoveUnusedProperties.java [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+
+        return new TreeVisitor<Tree, ExecutionContext>() {
+            @Override
+            public @Nullable Tree visit(@Nullable Tree tree, ExecutionContext ctx) {
+                if (tree instanceof SourceFile) {
+                    SourceFile sf = (SourceFile) tree;
+                    if (findPomUsagesVisitor.isAcceptable(sf, ctx)) { // ie: is a pom
+                        findPomUsagesVisitor.visit(sf, ctx);
+                        findFilteredResourcePathsVisitor.visit(sf, ctx);
+                    } else if (!(tree instanceof JavaSourceFile)) { // optimization: avoid visiting code files which are almost always not filtered resources
```
Comment (nmck257) [inline rewrite-maven/src/main/java/org/openrewrite/maven/RemoveUnusedProperties.java:120]: my first instinct was to check the `JavaSourceSet` marker for this, but I didn't see a way to differentiate between code files and resources with that marker

**Comment by nmck257** [comment]
fyi @timtebeek I'll probably merge this along later today, if no other concerns. I believe the added handling for filtered resources should be fairly robust

**Comment by timtebeek** [comment]
Apologies for the delay in reviewing this @nmck257 ; I'd been out since last Wednesday and only just now trying to catch up. I did just spot perhaps a small oversight: [the default delimiters for resource filtering](https://maven.apache.org/plugins/maven-resources-plugin/resources-mojo.html#delimiters) include `@a@` as well, which isn't picked up currently.

On my end to test this I have the Apache Maven repositories checked out with downloaded LSTs :
```
mod git clone moderne maven "Apache Maven"
mod build maven --download-only
```

Then once configured in the Moderne IntelliJ plugin:
image
It's fairly easy to preview the changes made to Apache Maven:
image

Ideally there's no false positives in such a run; whereas currently nearly all projects would see properties removed, which is perhaps a bit too enthusiastic (although not necessarily wrong).

Something to perhaps follow up, especially if you're similarly considering a run over there making changes with false positives.

**Pull request merged**